src/types/middleware.ts

```diff
@@ -0,0 +1,28 @@
+import { EffectScope } from 'vue'
+import { MaybePromise } from '@/types/utilities'
+
+type Route = {
+  name: string,
+  params: Record<string, unknown>,
+  query: unknown,
+  hash: string,
+}
+
+interface RouteState {
+
+}
+
+interface RouteRejection {
+  NotFound: unknown,
+}
+
+type RouteReject = (type?: keyof RouteRejection) => void
+
+type MiddleWareExtras = {
+  from: Route | null,
+  state: RouteState,
+  reject: RouteReject,
+  scope: EffectScope,
+}
+
+export type RouteMiddleware = (route: Route, extras: MiddleWareExtras) => MaybePromise<void>
```

Comment:
This is pretty simple stuff. The `route` is the route you're going to. `scope` is maybe not necessary but seemed like a nice to have. That will be the route's effect scope that is disposed when the route is exited.